test/grpc_test.rb

```diff
@@ -0,0 +1,215 @@
+      assert_equal :success, event
+      assert_equal Semian[:testing], resource
+      assert_equal :request_response, scope
+      assert_equal :grpc, adapter
+    end
+
+    run_services_on_server(@server, services: [EchoService]) do
+      GRPC::ActiveCall.any_instance.expects(:request_response)
+      @stub.an_rpc(EchoMsg.new)
+    end
+
+    assert notified, 'No notifications has been emitted'
+  ensure
+    Semian.unsubscribe(subscriber)
+  end
+
+  def test_circuit_breaker_on_client_streamer
+    stub = build_insecure_stub(EchoStub, host: "0.0.0.1:0", opts: {semian_options: SEMIAN_OPTIONS})
+    requests = [EchoMsg.new, EchoMsg.new]
+    open_circuit!(stub, :a_client_streaming_rpc, requests)
+
+    assert_raises GRPC::CircuitOpenError do
+      stub.a_client_streaming_rpc(requests)
+    end
+  end
+
+  def test_circuit_breaker_on_server_streamer
+    stub = build_insecure_stub(EchoStub, host: "0.0.0.1:0", opts: {semian_options: SEMIAN_OPTIONS})
+    request = EchoMsg.new
+    open_circuit!(stub, :a_server_streaming_rpc, request)
+
+    assert_raises GRPC::CircuitOpenError do
+      stub.a_server_streaming_rpc(request)
+    end
+  end
+
+  def test_circuit_breaker_on_bidi_streamer
+    stub = build_insecure_stub(EchoStub, host: "0.0.0.1:0", opts: {semian_options: SEMIAN_OPTIONS})
+    requests = [EchoMsg.new, EchoMsg.new]
+    open_circuit!(stub, :a_bidi_rpc, requests)
+
+    assert_raises GRPC::CircuitOpenError do
+      stub.a_bidi_rpc(requests)
+    end
+  end
+
+  def test_bulkheads_tickets_are_working
+    run_services_on_server(@server, services: [EchoService]) do
+      stub1 = build_insecure_stub(EchoStub, opts: {semian_options: SEMIAN_OPTIONS})
+      stub1.semian_resource.acquire do
+        testing2_semian_options = {
+          name: :testing2,
+          tickets: 1,
+          timeout: 0,
+          error_threshold: ERROR_THRESHOLD,
+          success_threshold: 1,
+          error_timeout: 10,
+        }
+        stub2 = build_insecure_stub(EchoStub, opts: {semian_options: testing2_semian_options})
+        stub2.semian_resource.acquire do
+          assert_raises GRPC::ResourceBusyError do
+            stub2.an_rpc(EchoMsg.new)
+          end
+        end
+      end
+    end
+  end
+
+  private
+
+  def open_circuit!(stub, method, args)
+    ERROR_THRESHOLD.times do
+      assert_raises GRPC::Unavailable do
+        stub.send(method, args)
+      end
+    end
+  end
+
+  def build_insecure_stub(klass, host: nil, opts: nil)
+    host ||= @host
+    opts ||= @client_opts
+    klass.new(host, :this_channel_is_insecure, **opts)
+  end
+
+  def build_rpc_server(server_opts: {}, client_opts: {})
+    @hostname = '0.0.0.0'
+    @server = new_rpc_server_for_testing({poll_period: 1}.merge(server_opts))
+    @port = @server.add_http2_port("#{@hostname}:0", :this_port_is_insecure)
+    @host = "#{@hostname}:#{@port}"
+    @client_opts = client_opts
+    populate_toxiproxy
+    @server
+  end
+
+  def populate_toxiproxy
```

Comment:
Why does this need to be dynamically populated vs using [hosts.yml.](https://github.com/Shopify/semian/blob/22e6dcf4d4425752a71db5e6284046cdc03fa1c4/test/config/hosts.yml)

Reply:
`@server.add_http2_port("#{@hostname}:0", :this_port_is_insecure)` returns a new port everytime I run it. I tried setting up a port directly but it does not work and had to populate toxiproxy on the go.

Reply:
Played with it a bit and I can use the `hosts.yml` but I am getting an error when trying to initialize the server in the test. `RuntimeError: could not add port 127.0.0.1:50051 to server, not sure why` I'll look into it

Reply:
Okay all good, fixed 👍 Thank you for the comment, forgot about that file!

Reply:
Awesome!